docs/architecture/adr-105-grpc-api.md

```diff
@@ -0,0 +1,222 @@
+
+## Context
+
+There has been discussion over the years as to which type of RPC interface would
+be preferable for Tendermint Core, and now CometBFT, to offer to integrators.
+[ADR 057][adr-057] captures some pros and cons of continuing to support the
+JSON-RPC API versus implementing a gRPC API. Previously it was decided to remove
+the gRPC API from Tendermint Core (see [tendermint/tendermint\#7121] and
+[tendermint/tendermint\#9683]).
+
+After discussion with users, and in considering the implementation of [ADR
+101][adr-101] (the data companion pull API), a decision has been taken to
+implement a gRPC API _in addition to_ the JSON-RPC API.
+
+The current plan is to implement this gRPC API after v0.38 after deprecating and
+subsequently removing the existing gRPC API (which only provides a
+`BroadcastService` with a single method). It is also envisaged that, once it is
+feasible to provide the RPC service independently of the node itself (see [ADR
+102][adr-102]), the JSON-RPC API on the node itself could eventually be
+deprecated and removed.
+
+## Alternative Approaches
+
+The primary alternative approach involves continuing to only provide support
+for, and potentially evolve, the JSON-RPC API. This API currently exposes many
+data types in non-standard and rather complex ways, making it difficult to
+implement clients. As per [ADR 075][adr-075], it also does not conform fully to
+the JSON-RPC 2.0 specification, further increasing client implementation
+complexity.
+
+## Decision
+
+Implement gRPC services corresponding to a minimal subset of the currently
+exposed [JSON-RPC endpoints][rpc-docs]. This set of services can always be
+expanded over time according to user needs, but once released it is hard to
+deprecate and remove such APIs.
+
+## Detailed Design
+
+### Services
+
+The initial services to be exposed via gRPC are informed by [Penumbra's
+`TendermintProxyService`][penumbra-proxy-svc], as well as the needs of the data
+companion API proposed in [ADR 101][adr-101]. Additional services can be rolled
+out additively in subsequent releases of CometBFT.
+
+Services are roughly organized by way of their respective domain. The details of
+each service, e.g. request/response types and precise Protobuf definitions, will
+be worked out in the implementation.
+
+- `VersionService` - A simple service that aims to be quite stable over time in
+  order to be utilized by clients to establish the version of the software with
+  which they are interacting (e.g. to pre-emptively determine compatibility).
+  This could technically be part of the `NodeService`, but if the `NodeService`
+  interface were to be modified, a new version of the service would need to be
+  created, and all old versions would need to be maintained, since the
+  `GetVersion` method needs to be quite stable.
+  - `GetVersion` - Query the version of the software and protocols employed by
+    the node (e.g. CometBFT, ABCI, block, P2P and application version).
```

Comment:
could add "version of any companions" to the list, even though it does not need to be exhaustive.

Reply:
I wouldn't. That would involve developing some more complex handshake mechanism with the data companion, for which I don't see a need at the moment.